### PR TITLE
Force xunit2 for pytest 6.0 when legacy mode is enabled (take III)

### PR DESCRIPTION
**Pull request by j-rivero**
Part III of my particular drama to implement correctly the check for pytest 6.0. The original implementation was in #415 and fixed after in #469. This last fix wrongly assumed that the pytest version check was happening inside the virtualenv. It was not detected by the CI jobs in the PR since the build node that run it happen to have pytest installed in the base system.

This PR tries to run the check inside the virtualenv on the testing job. First commit db9e19c back to enable the reverted code from yesterday, second commit are the real changes 8b6b63d I made to run on virtualenv.

Tested:
 * linux [Build Status](https://ci.ros2.org/job/ci_linux/11072/)
 * linux_aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/6386/)
 * win [Build Status](https://ci.ros2.org/job/ci_windows/10988/)
 * osx [Build Status](https://ci.ros2.org/job/ci_osx/9028/)
 * osx in mini1 (doesn't have pytest in base system) [Build Status](https://ci.ros2.org/job/ci_osx/9034/)

### CODE REVIEW

File: ros2_batch_job/__main__.py [hunk 1]
```diff
@@ -384,18 +384,27 @@ def build_and_test(args, job):
...
-            if config.get('pytest', 'junit_family') == 'xunit2':
+        if xunit_6_or_greater:
+            # only need to correct explicit legacy option if exists
+            if not check_xunit2_junit_family_value(config, 'legacy'):
                 continue
-        except configparser.NoOptionError:
-            pass
-        print('xunit2 patch applied to ' + str(path))
+        else:
+            # in xunit < 6 need to enforce xunit2 if not set
```
Comment (nuclearsandwich) [inline ros2_batch_job/__main__.py:404]: I think this is a typo. Isn't it the pytest version we're checking here?
```suggestion
            # in pytest < 6 need to enforce xunit2 if not set
```

File: ros2_batch_job/__main__.py [hunk 2]
```diff
@@ -384,18 +384,27 @@ def build_and_test(args, job):
     # xunit2 format is needed to make Jenkins xunit plugin 2.x happy
     with open('pytest.ini', 'w') as ini_file:
         ini_file.write('[pytest]\njunit_family=xunit2')
-    # check if packages have a pytest.ini file and add the xunit2
-    # format if it is not present
+    # check if packages have a pytest.ini file that doesn't choose junit_family=xunit2
+    # and patch configuration if needed to force the xunit2 value
+    xunit_6_or_greater = job.run([
```
Comment (nuclearsandwich) [inline ros2_batch_job/__main__.py:389]: I think this is a typo. Isn't it the pytest version we're checking here?
```suggestion
    pytest_6_or_greater = job.run([
```

File: ros2_batch_job/__main__.py [hunk 3]
```diff
@@ -384,18 +384,27 @@ def build_and_test(args, job):
...
+        'import sys;'
+        'sys.exit(StrictVersion(pytest.__version__) >= StrictVersion("6.0.0"))'
+        '\''])
     for path in Path('.').rglob('pytest.ini'):
         config = configparser.ConfigParser()
         config.read(str(path))
-        try:
-            # only if xunit2 is set continue the loop with the file unpatched
-            if config.get('pytest', 'junit_family') == 'xunit2':
+        if xunit_6_or_greater:
```
Comment (nuclearsandwich) [inline ros2_batch_job/__main__.py:399]: Fix variable name to match earlier suggestion.

```suggestion
        if pytest_6_or_greater:
```

**Comment by clalancette** [comment]
Can I make a suggestion here that might negate the need for this PR completely?

In order to have clean test runs, we need to install a pytest.ini file in all of the relevant repos.  Otherwise, pytest throws this xunit2 warning.  While this works around it, it still causes these warnings to show up when running locally.  I've been slowly adding pytest.ini to the relevant places as I have noticed them, but I might suggest that now would be a good opportunity to finish that work off.  Then we can remove this workaround from CI completely.

Thoughts?

**Comment by nuclearsandwich** [comment]
> I've been slowly adding pytest.ini to the relevant places as I have noticed them, but I might suggest that now would be a good opportunity to finish that work off. Then we can remove this workaround from CI completely.

As long as pytest 3.3.2 in Bionic supports the configuration option I'm completely totally 100% in favor of that approach. I was reviewing under the assumption that adding the appropriate configuration directly to affected repositories was off the table, otherwise why work around it.

If we can't backport the config option to Bionic it will affect Dashing and Eloquent CI runs on ci.ros2.org.

**Comment by clalancette** [comment]
Let me see what it will take to get rid of it for master.  Then I'll look at the options for backporting to bionic.

**Comment by nuclearsandwich** [comment]
@j-rivero if there is some reason we don't want to or can't make these changes at the source level can you add them to the PR description and let us know.

**Comment by clalancette** [comment]
All right, it looks doable.  It will take 19 PRs (that's just for master), plus probably an equivalent amount for Eloquent and Dashing, but I think that is a better way forward.  Once I get all of the PRs merged to master, I'll do a CI run with this workaround removed completely and see what it looks like.

**Comment by clalancette** [comment]
(also, as far as I know this should work fine for Ubuntu 18.04, so eloquent and dashing should be fine)

**Comment by j-rivero** [comment]
> @j-rivero if there is some reason we don't want to or can't make these changes at the source level can you add them to the PR description and let us know.

I [mentioned that briefly](https://github.com/ros2/ci/pull/406#issue-386759713) at the beginning of all the PRs serie. Nothing against it but that fix relies on no changes in current code with respect to how the format is configured in pytest.ini files and the same for new packages. If one of them is miss configured the whole build fails.

In order to be sure that the build is not going to fail in any case we can keep these changes in place together with fixes in source code level.

**Comment by dirk-thomas** [comment]
> If one of them is miss configured the whole build fails.

Why would the build fail? It only results in a warning is a package doesn't set the option and the workaround in CI is removed.

**Comment by dirk-thomas** [comment]
If all packages have the option defined we should revert the working in CI since there is a) no need for it and b) we want to be notified that any new package doesn't have option set (since a user building from source will experience the same).

**Comment by j-rivero** [comment]
> Why would the build fail? It only results in a warning is a package doesn't set the option and the workaround in CI is removed.

Would not be the same situation that leaded to https://github.com/ros2/ci/pull/393? The `junit_family` value is not the one expected by the xunit 2.x Jenkins plugin so the check on the format fails and the Jenkins plugin kills the whole build.

**Comment by dirk-thomas** [comment]
@j-rivero Can you please address the pending feedback so we can get this PR merged.

As a follow up we want to remove the workaround to generate custom `pytest.ini` files in CI for Rolling and forward.

**Comment by j-rivero** [comment]
> @j-rivero Can you please address the pending feedback so we can get this PR merged.

ready. A latest run to check these latest changes: [Build Status](https://ci.ros2.org/job/ci_windows/11314/)

**Comment by j-rivero** [comment]
>  A latest run to check these latest changes

probably would need a rebase or other alternatives to get changes from master in order to work.

**Comment by dirk-thomas** [comment]
@j-rivero What is the status on this ticket?

**Comment by j-rivero** [comment]
> @j-rivero What is the status on this ticket?

ready to merge as far as I can tell. I've merge master to get the final test build: [Build Status](https://ci.ros2.org/job/ci_windows/11669/)

File: ros2_batch_job/__main__.py [hunk 4]
```diff
@@ -359,18 +359,27 @@ def build_and_test(args, job):
     # xunit2 format is needed to make Jenkins xunit plugin 2.x happy
     with open('pytest.ini', 'w') as ini_file:
         ini_file.write('[pytest]\njunit_family=xunit2')
-    # check if packages have a pytest.ini file and add the xunit2
-    # format if it is not present
+    # check if packages have a pytest.ini file that doesn't choose junit_family=xunit2
+    # and patch configuration if needed to force the xunit2 value
+    pytest_6_or_greater = job.run([
+        '"%s"' % job.python, '-c', '\''
```
Comment (dirk-thomas) [inline ros2_batch_job/__main__.py:365]: Nitpick: assuming a double quote should work to use `'"'` instead of `'\''`.

Same for the closing quote.

Comment (j-rivero) [reply ros2_batch_job/__main__.py:365]: this can not use double quotes since they are being using in 369.

Comment (dirk-thomas) [reply ros2_batch_job/__main__.py:365]: Then avoid the escaping with `"'"`.

Comment (dirk-thomas) [reply ros2_batch_job/__main__.py:365]: Or use single quotes for the version string in line 369.

Comment (j-rivero) [reply ros2_batch_job/__main__.py]: thanks, I used `" ' "`

**Comment by j-rivero** [comment]
After merge with default and solve the quoting problem, last test run [Build Status](https://ci.ros2.org/job/ci_windows/11713/). Merging.

File: ros2_batch_job/__main__.py [hunk 5]
```diff
@@ -360,20 +360,30 @@ def build_and_test(args, job):
...
     # builds act the same.
     if args.ros_distro in ('dashing', 'eloquent', 'foxy'):
+        # xunit2 format is needed to make Jenkins xunit plugin 2.x happy
         with open('pytest.ini', 'w') as ini_file:
             ini_file.write('[pytest]\njunit_family=xunit2')
-        # check if packages have a pytest.ini file and add the xunit2
-        # format if it is not present
+        # check if packages have a pytest.ini file that doesn't choose junit_family=xunit2
+        # and patch configuration if needed to force the xunit2 value
+        pytest_6_or_greater = job.run([
```
Comment (jacobperron) [inline ros2_batch_job/__main__.py:368]: Looks like if this check fails, it fails the build ([for example](https://ci.ros2.org/job/ci_linux/11784/)).

Comment (jacobperron) [reply ros2_batch_job/__main__.py:368]: See https://github.com/ros2/ci/pull/498 for a proposed fix.

**Pull request merged**